File: pkg/client/journal.go
package client

import (
	"bufio"
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strconv"
	"strings"

	"github.com/mem9-ai/dat9/pkg/journal"
)

func (c *Client) CreateJournal(ctx context.Context, req journal.CreateRequest) (*journal.Journal, error) {
	body, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, c.baseURL+"/v1/journals", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	httpReq.Header.Set("Content-Type", "application/json")
	resp, err := c.do(httpReq)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= 300 {
		return nil, readError(resp)
	}
	var out journal.Journal
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("decode journal create: %w", err)
	}
	return &out, nil
}

func (c *Client) AppendJournalEntries(ctx context.Context, journalID, appendID string, entries []journal.EntryInput) (*journal.AppendResponse, error) {
	body, err := json.Marshal(entries)
	if err != nil {
		return nil, err
	}
	u := c.baseURL + "/v1/journals/" + url.PathEscape(journalID) + "/entries"
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, u, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Idempotency-Key", appendID)
	resp, err := c.do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= 300 {
		return nil, readError(resp)
	}
	var out journal.AppendResponse
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("decode journal append: %w", err)
	}
	return &out, nil
}

func (c *Client) ReadJournalEntries(ctx context.Context, journalID string, afterSeq int64, limit int) ([]journal.Entry, error) {
	values := url.Values{}
	if afterSeq > 0 {
		values.Set("after_seq", strconv.FormatInt(afterSeq, 10))
	}
	if limit > 0 {
		values.Set("limit", strconv.Itoa(limit))
	}
	u := c.baseURL + "/v1/journals/" + url.PathEscape(journalID) + "/entries"
	if encoded := values.Encode(); encoded != "" {
		u += "?" + encoded
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= 300 {
		return nil, readError(resp)
	}
	var out []journal.Entry
	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 64*1024), 4<<20)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var entry journal.Entry
		if err := json.Unmarshal([]byte(line), &entry); err != nil {
			return nil, fmt.Errorf("decode journal entry: %w", err)
		}
		out = append(out, entry)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return out, nil
}

func (c *Client) SearchJournal(ctx context.Context, req journal.SearchRequest) ([]journal.SearchMatch, error) {
	values := url.Values{}
	if req.Type != "" {
		values.Set("type", req.Type)
	}
	if req.Status != "" {
		values.Set("status", req.Status)
	}
	if req.Kind != "" {
		values.Set("kind", req.Kind)
	}
	if req.ActorType != "" {
		values.Set("actor", req.ActorType+":"+req.ActorID)
	}
	for _, subject := range req.Subjects {
		values.Add("subject", subject)
	}
	for _, label := range journal.NormalizeLabels(req.Labels) {
		values.Add("meta", label.Key+"="+label.Value)
	}
	if req.SinceRaw != "" {
		values.Set("since", req.SinceRaw)
	} else if req.Since != nil {
		values.Set("since", journal.FormatTime(*req.Since))
	}
	if req.UntilRaw != "" {
		values.Set("until", req.UntilRaw)
	} else if req.Until != nil {
		values.Set("until", journal.FormatTime(*req.Until))
	}
	if req.Limit > 0 {
		values.Set("limit", strconv.Itoa(req.Limit))
	}
	if req.Cursor != "" {
		values.Set("cursor", req.Cursor)
	}
	if req.Entries {
		values.Set("include", "entry")
	}
	u := c.baseURL + "/v1/journal-entries"
	if encoded := values.Encode(); encoded != "" {
		u += "?" + encoded
	}
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.do(httpReq)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= 300 {
		return nil, readError(resp)
	}
	var out []journal.SearchMatch
	scanner := bufio.NewScanner(resp.Body)
	scanner.Buffer(make([]byte, 64*1024), 4<<20)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}
		var match journal.SearchMatch
		if err := json.Unmarshal([]byte(line), &match); err != nil {
			return nil, fmt.Errorf("decode journal match: %w", err)
		}
		out = append(out, match)
	}
	if err := scanner.Err(); err != nil {
		return nil, err
	}
	return out, nil
}

func (c *Client) VerifyJournal(ctx context.Context, journalID string) (*journal.VerifyResult, error) {
	u := c.baseURL + "/v1/journals/" + url.PathEscape(journalID) + "/verify"
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, u, nil)
	if err != nil {
		return nil, err
	}
	resp, err := c.do(req)
	if err != nil {
		return nil, err
	}
	defer func() { _ = resp.Body.Close() }()
	if resp.StatusCode >= 300 {
		return nil, readError(resp)
	}
	var out journal.VerifyResult
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return nil, fmt.Errorf("decode journal verify: %w", err)
	}
	return &out, nil
}

File: clients/drive9-js/src/journal.ts
import type { Client } from "./client.js";
import { checkError } from "./error.js";
import type {
  Journal,
  JournalAppendResponse,
  JournalCreateRequest,
  JournalEntry,
  JournalEntryInput,
  JournalSearchMatch,
  JournalSearchRequest,
  JournalVerifyResult,
} from "./models.js";

async function json<T>(resp: Response): Promise<T> {
  await checkError(resp);
  return (await resp.json()) as T;
}

async function ndjson<T>(resp: Response): Promise<T[]> {
  await checkError(resp);
  const text = await resp.text();
  return text
    .split(/\r?\n/)
    .map((line) => line.trim())
    .filter(Boolean)
    .map((line) => JSON.parse(line) as T);
}

export async function createJournal(client: Client, req: JournalCreateRequest): Promise<Journal> {
  return json<Journal>(
    await fetch(`${client.baseUrl}/v1/journals`, {
      method: "POST",
      headers: client.authHeaders({ "Content-Type": "application/json" }),
      body: JSON.stringify(req),
    })
  );
}

export async function appendJournalEntries(
  client: Client,
  journalId: string,
  appendId: string,
  entries: JournalEntryInput[]
): Promise<JournalAppendResponse> {
  return json<JournalAppendResponse>(
    await fetch(`${client.baseUrl}/v1/journals/${encodeURIComponent(journalId)}/entries`, {
      method: "POST",
      headers: client.authHeaders({
        "Content-Type": "application/json",
        "Idempotency-Key": appendId,
      }),
      body: JSON.stringify(entries),
    })
  );
}

export async function readJournalEntries(
  client: Client,
  journalId: string,
  afterSeq = 0,
  limit = 0
): Promise<JournalEntry[]> {
  const qs = new URLSearchParams();
  if (afterSeq > 0) qs.set("after_seq", String(afterSeq));
  if (limit > 0) qs.set("limit", String(limit));
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return ndjson<JournalEntry>(
    await fetch(`${client.baseUrl}/v1/journals/${encodeURIComponent(journalId)}/entries${suffix}`, {
      headers: client.authHeaders(),
    })
  );
}

export async function searchJournal(client: Client, req: JournalSearchRequest): Promise<JournalSearchMatch[]> {
  const qs = new URLSearchParams();
  if (req.type) qs.set("type", req.type);
  if (req.status) qs.set("status", req.status);
  if (req.kind) qs.set("kind", req.kind);
  if (req.actor_type) qs.set("actor", `${req.actor_type}:${req.actor_id || ""}`);
  for (const subject of req.subjects || []) qs.append("subject", subject);
  for (const label of req.labels || []) qs.append("meta", `${label.key}=${label.value}`);
  if (req.since) qs.set("since", req.since);
  if (req.until) qs.set("until", req.until);
  if (req.limit && req.limit > 0) qs.set("limit", String(req.limit));
  if (req.cursor) qs.set("cursor", req.cursor);
  if (req.entries) qs.set("include", "entry");
  const suffix = qs.toString() ? `?${qs.toString()}` : "";
  return ndjson<JournalSearchMatch>(
    await fetch(`${client.baseUrl}/v1/journal-entries${suffix}`, { headers: client.authHeaders() })
  );
}

export async function verifyJournal(client: Client, journalId: string): Promise<JournalVerifyResult> {
  return json<JournalVerifyResult>(
    await fetch(`${client.baseUrl}/v1/journals/${encodeURIComponent(journalId)}/verify`, {
      headers: client.authHeaders(),
    })
  );
}
